# Exploratory Data Analysis MIMIC III

In [1]:
import sqlite3
import pandas as pd

pd.set_option('display.max_colwidth', None)  # Show full content of columns
pd.set_option('display.max_rows', None)      # Show all rows

# Determine DB schema

In [2]:
# Connect to sql db 
conn = sqlite3.connect("../data/mimic.db")

# Determine what tables Exist 
pd.read_sql(
    """
    SELECT name 
    FROM sqlite_master
    WHERE type='table' 
    ORDER BY name
    """,
    conn
)

,name
0,admissions
1,callout
2,caregivers
3,chartevents
4,cptevents
5,d_icd_diagnoses
6,d_icd_procedures
7,d_items
8,d_labitems
9,datetimeevents


In [33]:
# As it can be tedious to write all the names this is a query generator. Now we just copy paste into the actual query we want
print(pd.read_sql(
    """
    SELECT 'SELECT ''' || name || ''' AS table_name, COUNT(*) AS row_count FROM ' || name || ' UNION ALL'
    FROM sqlite_master 
    WHERE type='table'; 
    """, 
    conn
))

        'SELECT ''' || name || ''' AS table_name, COUNT(*) AS row_count FROM ' || name || ' UNION ALL'
0                       SELECT 'patients' AS table_name, COUNT(*) AS row_count FROM patients UNION ALL
1                   SELECT 'admissions' AS table_name, COUNT(*) AS row_count FROM admissions UNION ALL
2                       SELECT 'icustays' AS table_name, COUNT(*) AS row_count FROM icustays UNION ALL
3                 SELECT 'chartevents' AS table_name, COUNT(*) AS row_count FROM chartevents UNION ALL
4                     SELECT 'labevents' AS table_name, COUNT(*) AS row_count FROM labevents UNION ALL
5               SELECT 'outputevents' AS table_name, COUNT(*) AS row_count FROM outputevents UNION ALL
6           SELECT 'inputevents_cv' AS table_name, COUNT(*) AS row_count FROM inputevents_cv UNION ALL
7           SELECT 'inputevents_mv' AS table_name, COUNT(*) AS row_count FROM inputevents_mv UNION ALL
8             SELECT 'prescriptions' AS table_name, COUNT(*) AS row_count

In [34]:
# View row counts of each Table (Should only be 100 Patients as its the DEMO Version of MIMIC 3)
pd.read_sql(
    """
    SELECT 'patients' AS table_name, COUNT(*) AS row_count FROM patients UNION ALL
    SELECT 'admissions' AS table_name, COUNT(*) AS row_count FROM admissions UNION ALL
    SELECT 'icustays' AS table_name, COUNT(*) AS row_count FROM icustays UNION ALL
    SELECT 'chartevents' AS table_name, COUNT(*) AS row_count FROM chartevents UNION ALL
    SELECT 'labevents' AS table_name, COUNT(*) AS row_count FROM labevents UNION ALL
    SELECT 'outputevents' AS table_name, COUNT(*) AS row_count FROM outputevents UNION ALL
    SELECT 'inputevents_cv' AS table_name, COUNT(*) AS row_count FROM inputevents_cv UNION ALL
    SELECT 'inputevents_mv' AS table_name, COUNT(*) AS row_count FROM inputevents_mv UNION ALL
    SELECT 'prescriptions' AS table_name, COUNT(*) AS row_count FROM prescriptions UNION ALL
    SELECT 'procedures_icd' AS table_name, COUNT(*) AS row_count FROM procedures_icd UNION ALL
    SELECT 'diagnoses_icd' AS table_name, COUNT(*) AS row_count FROM diagnoses_icd UNION ALL
    SELECT 'd_icd_diagnoses' AS table_name, COUNT(*) AS row_count FROM d_icd_diagnoses UNION ALL
    SELECT 'd_icd_procedures' AS table_name, COUNT(*) AS row_count FROM d_icd_procedures UNION ALL
    SELECT 'd_items' AS table_name, COUNT(*) AS row_count FROM d_items UNION ALL
    SELECT 'd_labitems' AS table_name, COUNT(*) AS row_count FROM d_labitems UNION ALL
    SELECT 'microbiologyevents' AS table_name, COUNT(*) AS row_count FROM microbiologyevents UNION ALL
    SELECT 'datetimeevents' AS table_name, COUNT(*) AS row_count FROM datetimeevents UNION ALL
    SELECT 'callout' AS table_name, COUNT(*) AS row_count FROM callout UNION ALL
    SELECT 'caregivers' AS table_name, COUNT(*) AS row_count FROM caregivers UNION ALL
    SELECT 'services' AS table_name, COUNT(*) AS row_count FROM services UNION ALL
    SELECT 'transfers' AS table_name, COUNT(*) AS row_count FROM transfers UNION ALL
    SELECT 'cptevents' AS table_name, COUNT(*) AS row_count FROM cptevents UNION ALL
    SELECT 'drgcodes' AS table_name, COUNT(*) AS row_count FROM drgcodes
    """,
    conn
)

,table_name,row_count
0,patients,100
1,admissions,129
2,icustays,136
3,chartevents,758355
4,labevents,76074
5,outputevents,11320
6,inputevents_cv,34799
7,inputevents_mv,13224
8,prescriptions,10398
9,procedures_icd,506


In [35]:
# View each table for 5 rows 
tables = conn.execute(
    "SELECT name FROM sqlite_master WHERE type='table' ORDER BY name"
).fetchall()

for (t,) in tables:
    print(f"\n{'='*60}")
    print(f" {t.upper()}")
    print(f"{'='*60}")
    display(pd.read_sql(f"SELECT * FROM {t} LIMIT 5", conn))


 ADMISSIONS


,row_id,subject_id,hadm_id,admittime,dischtime,deathtime,admission_type,admission_location,discharge_location,insurance,language,religion,marital_status,ethnicity,edregtime,edouttime,diagnosis,hospital_expire_flag,has_chartevents_data
0,12258,10006,142345,2164-10-23 21:09:00,2164-11-01 17:15:00,None,EMERGENCY,EMERGENCY ROOM ADMIT,HOME HEALTH CARE,Medicare,None,CATHOLIC,SEPARATED,BLACK/AFRICAN AMERICAN,2164-10-23 16:43:00,2164-10-23 23:00:00,SEPSIS,0,1
1,12263,10011,105331,2126-08-14 22:32:00,2126-08-28 18:59:00,2126-08-28 18:59:00,EMERGENCY,TRANSFER FROM HOSP/EXTRAM,DEAD/EXPIRED,Private,None,CATHOLIC,SINGLE,UNKNOWN/NOT SPECIFIED,None,None,HEPATITIS B,1,1
2,12265,10013,165520,2125-10-04 23:36:00,2125-10-07 15:13:00,2125-10-07 15:13:00,EMERGENCY,TRANSFER FROM HOSP/EXTRAM,DEAD/EXPIRED,Medicare,None,CATHOLIC,None,UNKNOWN/NOT SPECIFIED,None,None,SEPSIS,1,1
3,12269,10017,199207,2149-05-26 17:19:00,2149-06-03 18:42:00,None,EMERGENCY,EMERGENCY ROOM ADMIT,SNF,Medicare,None,CATHOLIC,DIVORCED,WHITE,2149-05-26 12:08:00,2149-05-26 19:45:00,HUMERAL FRACTURE,0,1
4,12270,10019,177759,2163-05-14 20:43:00,2163-05-15 12:00:00,2163-05-15 12:00:00,EMERGENCY,TRANSFER FROM HOSP/EXTRAM,DEAD/EXPIRED,Medicare,None,CATHOLIC,DIVORCED,WHITE,None,None,ALCOHOLIC HEPATITIS,1,1



 CALLOUT


,row_id,subject_id,hadm_id,submit_wardid,submit_careunit,curr_wardid,curr_careunit,callout_wardid,callout_service,request_tele,...,callout_status,callout_outcome,discharge_wardid,acknowledge_status,createtime,updatetime,acknowledgetime,outcometime,firstreservationtime,currentreservationtime
0,3917,10017,199207,7,None,45,CCU,1,MED,1,...,Inactive,Discharged,45.0,Acknowledged,2149-05-31 10:44:34,2149-05-31 10:44:34,2149-05-31 15:08:04,2149-05-31 22:40:02,None,None
1,3919,10026,103770,33,None,3,SICU,3,NMED,1,...,Inactive,Discharged,3.0,Revised,2195-05-18 13:56:20,2195-05-19 15:45:30,None,2195-05-19 17:40:03,None,None
2,3920,10027,199395,12,None,55,CSRU,55,CSURG,1,...,Inactive,Discharged,55.0,Acknowledged,2190-07-20 08:15:20,2190-07-20 08:15:20,2190-07-20 08:57:46,2190-07-20 17:10:02,None,None
3,3921,10029,132349,33,None,45,SICU,1,MED,0,...,Inactive,Discharged,45.0,Acknowledged,2139-09-24 09:53:37,2139-09-24 09:53:37,2139-09-24 09:56:02,2139-09-25 19:10:01,None,None
4,3922,10033,157235,33,None,4,SICU,1,MED,1,...,Inactive,Discharged,4.0,Revised,2132-12-06 10:16:08,2132-12-06 14:53:53,None,2132-12-06 15:10:02,None,None



 CAREGIVERS


,row_id,cgid,label,description
0,2228,16174,RO,Read Only
1,2229,16175,RO,Read Only
2,2230,16176,Res,Resident/Fellow/PA/NP
3,2231,16177,RO,Read Only
4,2232,16178,RT,Respiratory



 CHARTEVENTS


,row_id,subject_id,hadm_id,icustay_id,itemid,charttime,storetime,cgid,value,valuenum,valueuom,warning,error,resultstatus,stopped
0,5279021,40124,126179,279554.0,223761,2130-02-04 04:00:00,2130-02-04 04:35:00,19085,95.9,95.9,?F,0,0,None,None
1,5279022,40124,126179,279554.0,224695,2130-02-04 04:25:00,2130-02-04 05:55:00,18999,2222221.7,2222221.7,cmH2O,0,0,None,None
2,5279023,40124,126179,279554.0,220210,2130-02-04 04:30:00,2130-02-04 04:43:00,21452,15.0,15.0,insp/min,0,0,None,None
3,5279024,40124,126179,279554.0,220045,2130-02-04 04:32:00,2130-02-04 04:43:00,21452,94.0,94.0,bpm,0,0,None,None
4,5279025,40124,126179,279554.0,220179,2130-02-04 04:32:00,2130-02-04 04:43:00,21452,163.0,163.0,mmHg,0,0,None,None



 CPTEVENTS


,row_id,subject_id,hadm_id,costcenter,chartdate,cpt_cd,cpt_number,cpt_suffix,ticket_id_seq,sectionheader,subsectionheader,description
0,4615,10117,105150,ICU,None,99254,99254,None,1.0,Evaluation and management,Consultations,None
1,4616,10117,105150,ICU,None,99231,99231,None,2.0,Evaluation and management,Hospital inpatient services,None
2,4617,10117,105150,ICU,None,90935,90935,None,3.0,Medicine,Dialysis,None
3,4618,10117,105150,ICU,None,99231,99231,None,4.0,Evaluation and management,Hospital inpatient services,None
4,7753,10111,174739,ICU,None,99253,99253,None,1.0,Evaluation and management,Consultations,None



 D_ICD_DIAGNOSES


,row_id,icd9_code,short_title,long_title
0,1,01716,Erythem nod tb-oth test,"Erythema nodosum with hypersensitivity reaction in tuberculosis, tubercle bacilli not found by bacteriological or histological examination, but tuberculosis confirmed by other methods [inoculation of animals]"
1,2,01720,TB periph lymph-unspec,"Tuberculosis of peripheral lymph nodes, unspecified"
2,3,01721,TB periph lymph-no exam,"Tuberculosis of peripheral lymph nodes, bacteriological or histological examination not done"
3,4,01722,TB periph lymph-exam unk,"Tuberculosis of peripheral lymph nodes, bacteriological or histological examination unknown (at present)"
4,5,01723,TB periph lymph-micro dx,"Tuberculosis of peripheral lymph nodes, tubercle bacilli found (in sputum) by microscopy"



 D_ICD_PROCEDURES


,row_id,icd9_code,short_title,long_title
0,1,1423,Chorioret les xenon coag,Destruction of chorioretinal lesion by xenon arc photocoagulation
1,2,1424,Chorioret les laser coag,Destruction of chorioretinal lesion by laser photocoagulation
2,3,1425,Chorioret les p/coag NOS,Destruction of chorioretinal lesion by photocoagulation of unspecified type
3,4,1426,Chorioret les radiother,Destruction of chorioretinal lesion by radiation therapy
4,5,1427,Chorioret les rad implan,Destruction of chorioretinal lesion by implantation of radiation source



 D_ITEMS


,row_id,itemid,label,abbreviation,dbsource,linksto,category,unitname,param_type,conceptid
0,1,1435,Sustained Nystamus,None,carevue,chartevents,None,None,None,None
1,2,1436,Tactile Disturbances,None,carevue,chartevents,None,None,None,None
2,3,1437,Tremor,None,carevue,chartevents,None,None,None,None
3,4,1438,Ulnar Pulse [Right],None,carevue,chartevents,None,None,None,None
4,5,1439,Visual Disturbances,None,carevue,chartevents,None,None,None,None



 D_LABITEMS


,row_id,itemid,label,fluid,category,loinc_code
0,1,50800,SPECIMEN TYPE,BLOOD,BLOOD GAS,None
1,2,50801,Alveolar-arterial Gradient,Blood,Blood Gas,19991-9
2,3,50802,Base Excess,Blood,Blood Gas,11555-0
3,4,50803,"Calculated Bicarbonate, Whole Blood",Blood,Blood Gas,1959-6
4,5,50804,Calculated Total CO2,Blood,Blood Gas,34728-6



 DATETIMEEVENTS


,row_id,subject_id,hadm_id,icustay_id,itemid,charttime,storetime,cgid,value,valueuom,warning,error,resultstatus,stopped
0,208474,10076,198503,201006.0,5684,2107-03-25 04:00:00,2107-03-25 04:34:00,20482,2107-03-24 00:00:00,Date,None,None,None,NotStopd
1,208475,10076,198503,201006.0,5684,2107-03-25 07:00:00,2107-03-25 07:06:00,15004,2107-03-24 00:00:00,Date,None,None,None,NotStopd
2,208836,10076,198503,201006.0,5684,2107-03-26 04:00:00,2107-03-26 05:31:00,20834,2107-03-24 00:00:00,Date,None,None,None,NotStopd
3,208837,10076,198503,201006.0,5684,2107-03-26 08:00:00,2107-03-26 08:33:00,17480,2107-03-24 00:00:00,Date,None,None,None,NotStopd
4,208838,10076,198503,201006.0,5684,2107-03-26 16:00:00,2107-03-26 16:08:00,17480,2107-03-24 00:00:00,Date,None,None,None,NotStopd



 DIAGNOSES_ICD


,row_id,subject_id,hadm_id,seq_num,icd9_code
0,112344,10006,142345,1,99591
1,112345,10006,142345,2,99662
2,112346,10006,142345,3,5672
3,112347,10006,142345,4,40391
4,112348,10006,142345,5,42731



 DRGCODES


,row_id,subject_id,hadm_id,drg_type,drg_code,description,drg_severity,drg_mortality
0,1338,10130,156668,HCFA,148,"MAJOR SMALL & LARGE BOWEL PROCEDURES WITH COMPLICATIONS, COMORBIDITIES",None,None
1,2188,10114,167957,HCFA,518,"PERCUTANEOUS CARDIOVASCULAR PROCEDURES WITHOUT ACUTE MYOCARDIAL INFARCTION, WITHOUT CORONARY ARTERY STENT IMPLANT",None,None
2,2599,10117,187023,HCFA,185,DENTAL & ORAL DIS EXCEPT EXTRACTIONS & RESTORATIONS AGE >17,None,None
3,2703,10046,133110,HCFA,1,CRANIOTOMY AGE >17 EXCEPT FOR TRAUMA,None,None
4,3020,10011,105331,HCFA,205,"DISORDERS OF LIVER EXCEPT MALIGNANCY, CIRRHOSIS, ALCOHOLIC HEPATITIS WITH COMPLICATIONS, COMORBIDITIES",None,None



 ICUSTAYS


,row_id,subject_id,hadm_id,icustay_id,dbsource,first_careunit,last_careunit,first_wardid,last_wardid,intime,outtime,los
0,12742,10006,142345,206504,carevue,MICU,MICU,52,52,2164-10-23 21:10:15,2164-10-25 12:21:07,1.6325
1,12747,10011,105331,232110,carevue,MICU,MICU,15,15,2126-08-14 22:34:00,2126-08-28 18:59:00,13.8507
2,12749,10013,165520,264446,carevue,MICU,MICU,15,15,2125-10-04 23:38:00,2125-10-07 15:13:52,2.6499
3,12754,10017,199207,204881,carevue,CCU,CCU,7,7,2149-05-29 18:52:29,2149-05-31 22:19:17,2.1436
4,12755,10019,177759,228977,carevue,MICU,MICU,15,15,2163-05-14 20:43:56,2163-05-16 03:47:04,1.2938



 INPUTEVENTS_CV


,row_id,subject_id,hadm_id,icustay_id,charttime,itemid,amount,amountuom,rate,rateuom,...,orderid,linkorderid,stopped,newbottle,originalamount,originalamountuom,originalroute,originalrate,originalrateuom,originalsite
0,1184,10114,167957,234989,2171-11-03 15:00:00,30056,400.0,ml,None,None,...,2557279,2557279,None,None,None,ml,Oral,None,None,None
1,1185,10114,167957,234989,2171-11-03 20:00:00,30056,120.0,ml,None,None,...,7828849,2557279,None,None,None,ml,Oral,None,None,None
2,1186,10114,167957,234989,2171-11-03 23:00:00,30056,120.0,ml,None,None,...,2744159,2557279,None,None,None,ml,Oral,None,None,None
3,1187,10114,167957,234989,2171-11-04 02:00:00,30056,120.0,ml,None,None,...,8475006,2557279,None,None,None,ml,Oral,None,None,None
4,1188,10114,167957,234989,2171-11-04 05:00:00,30056,120.0,ml,None,None,...,11183474,2557279,None,None,None,ml,Oral,None,None,None



 INPUTEVENTS_MV


,row_id,subject_id,hadm_id,icustay_id,starttime,endtime,itemid,amount,amountuom,rate,...,totalamountuom,isopenbag,continueinnextdept,cancelreason,statusdescription,comments_editedby,comments_canceledby,comments_date,originalamount,originalrate
0,118897,42367,139932,250305,2147-10-29 16:45:00,2147-10-29 16:46:00,225799,60.000000,ml,NaN,...,ml,0,0,0,FinishedRunning,None,None,None,60.0,60.000000
1,118898,42367,139932,250305,2147-10-20 13:17:00,2147-10-20 13:18:00,223258,10.000000,units,NaN,...,None,0,0,1,Rewritten,None,RN,2147-10-20 13:18:00,10.0,10.000000
2,118899,42367,139932,250305,2147-10-29 03:23:00,2147-10-29 03:53:00,226089,99.999999,ml,199.999998,...,ml,0,0,0,FinishedRunning,None,None,None,100.0,200.000000
3,118900,42367,139932,250305,2147-10-22 22:00:00,2147-10-22 22:01:00,225799,40.000000,ml,NaN,...,ml,0,0,0,FinishedRunning,None,None,None,40.0,40.000000
4,118901,42367,139932,250305,2147-10-16 06:21:00,2147-10-17 06:10:00,225936,1309.899995,ml,54.999300,...,ml,0,0,0,FinishedRunning,None,None,None,1309.9,54.999298



 LABEVENTS


,row_id,subject_id,hadm_id,itemid,charttime,value,valuenum,valueuom,flag
0,6244563,10006,None,50868,2164-09-24 20:21:00,19,19.0,mEq/L,None
1,6244564,10006,None,50882,2164-09-24 20:21:00,27,27.0,mEq/L,None
2,6244565,10006,None,50893,2164-09-24 20:21:00,10.0,10.0,mg/dL,None
3,6244566,10006,None,50902,2164-09-24 20:21:00,97,97.0,mEq/L,None
4,6244567,10006,None,50912,2164-09-24 20:21:00,7.0,7.0,mg/dL,abnormal



 MICROBIOLOGYEVENTS


,row_id,subject_id,hadm_id,chartdate,charttime,spec_itemid,spec_type_desc,org_itemid,org_name,isolate_num,ab_itemid,ab_name,dilution_text,dilution_comparison,dilution_value,interpretation
0,134694,10006,142345,2164-10-23,2164-10-23 15:30:00,70012,BLOOD CULTURE,80155.0,"STAPHYLOCOCCUS, COAGULASE NEGATIVE",2.0,NaN,None,None,None,NaN,None
1,134695,10006,142345,2164-10-23,2164-10-23 15:30:00,70012,BLOOD CULTURE,80155.0,"STAPHYLOCOCCUS, COAGULASE NEGATIVE",1.0,90015.0,VANCOMYCIN,2,=,2.0,S
2,134696,10006,142345,2164-10-23,2164-10-23 15:30:00,70012,BLOOD CULTURE,80155.0,"STAPHYLOCOCCUS, COAGULASE NEGATIVE",1.0,90012.0,GENTAMICIN,<=0.5,<=,1.0,S
3,134697,10006,142345,2164-10-23,2164-10-23 15:30:00,70012,BLOOD CULTURE,80155.0,"STAPHYLOCOCCUS, COAGULASE NEGATIVE",1.0,90025.0,LEVOFLOXACIN,4,=,4.0,I
4,134698,10006,142345,2164-10-23,2164-10-23 15:30:00,70012,BLOOD CULTURE,80155.0,"STAPHYLOCOCCUS, COAGULASE NEGATIVE",1.0,90016.0,OXACILLIN,=>4,=>,4.0,R



 OUTPUTEVENTS


,row_id,subject_id,hadm_id,icustay_id,charttime,itemid,value,valueuom,storetime,cgid,stopped,newbottle,iserror
0,6540,10114,167957,234989.0,2171-10-30 20:00:00,40055,39.0,ml,2171-10-30 20:38:00,15029,None,None,None
1,6541,10114,167957,234989.0,2171-10-30 21:00:00,40055,35.0,ml,2171-10-30 21:18:00,15029,None,None,None
2,6542,10114,167957,234989.0,2171-10-30 23:00:00,40055,100.0,ml,2171-10-30 23:31:00,15029,None,None,None
3,6543,10114,167957,234989.0,2171-10-31 00:00:00,40055,45.0,ml,2171-10-31 00:24:00,15029,None,None,None
4,6544,10114,167957,234989.0,2171-10-31 02:00:00,40055,80.0,ml,2171-10-31 02:02:00,15029,None,None,None



 PATIENTS


,row_id,subject_id,gender,dob,dod,dod_hosp,dod_ssn,expire_flag
0,9467,10006,F,2094-03-05,2165-08-12,2165-08-12,2165-08-12,1
1,9472,10011,F,2090-06-05,2126-08-28,2126-08-28,None,1
2,9474,10013,F,2038-09-03,2125-10-07,2125-10-07,2125-10-07,1
3,9478,10017,F,2075-09-21,2152-09-12,None,2152-09-12,1
4,9479,10019,M,2114-06-20,2163-05-15,2163-05-15,2163-05-15,1



 PRESCRIPTIONS


,row_id,subject_id,hadm_id,icustay_id,startdate,enddate,drug_type,drug,drug_name_poe,drug_name_generic,formulary_drug_cd,gsn,ndc,prod_strength,dose_val_rx,dose_unit_rx,form_val_disp,form_unit_disp,route
0,32600,42458,159647,None,2146-07-21,2146-07-22,MAIN,Pneumococcal Vac Polyvalent,Pneumococcal Vac Polyvalent,PNEUMOcoccal Vac Polyvalent,PNEU25I,48548.0,6494300.0,25mcg/0.5mL Vial,0.5,mL,1,VIAL,IM
1,32601,42458,159647,None,2146-07-21,2146-07-22,MAIN,Bisacodyl,Bisacodyl,Bisacodyl,BISA5,2947.0,536338101.0,5 mg Tab,10,mg,2,TAB,PO
2,32602,42458,159647,None,2146-07-21,2146-07-22,MAIN,Bisacodyl,Bisacodyl,Bisacodyl (Rectal),BISA10R,2944.0,574705050.0,10mg Suppository,10,mg,1,SUPP,PR
3,32603,42458,159647,None,2146-07-21,2146-07-22,MAIN,Senna,Senna,Senna,SENN187,19964.0,904516561.0,1 Tablet,1,TAB,1,TAB,PO
4,32604,42458,159647,None,2146-07-21,2146-07-21,MAIN,Docusate Sodium (Liquid),Docusate Sodium (Liquid),Docusate Sodium (Liquid),DOCU100L,3017.0,121054410.0,100mg UD Cup,100,mg,1,UDCUP,PO



 PROCEDURES_ICD


,row_id,subject_id,hadm_id,seq_num,icd9_code
0,3994,10114,167957,1,3605
1,3995,10114,167957,2,3722
2,3996,10114,167957,3,8856
3,3997,10114,167957,4,9920
4,3998,10114,167957,5,9671



 SERVICES


,row_id,subject_id,hadm_id,transfertime,prev_service,curr_service
0,14974,10006,142345,2164-10-23 21:10:15,None,MED
1,14979,10011,105331,2126-08-14 22:34:00,None,MED
2,14981,10013,165520,2125-10-04 23:38:00,None,MED
3,14985,10017,199207,2149-05-26 17:21:09,None,MED
4,14986,10019,177759,2163-05-14 20:43:56,None,MED



 TRANSFERS


,row_id,subject_id,hadm_id,icustay_id,dbsource,eventtype,prev_careunit,curr_careunit,prev_wardid,curr_wardid,intime,outtime,los
0,54440,10006,142345,206504.0,carevue,admit,None,MICU,NaN,52.0,2164-10-23 21:10:15,2164-10-25 12:21:07,39.18
1,54441,10006,142345,NaN,carevue,transfer,MICU,None,52.0,45.0,2164-10-25 12:21:07,2164-11-01 17:14:27,172.89
2,54442,10006,142345,NaN,carevue,discharge,None,None,45.0,NaN,2164-11-01 17:14:27,None,NaN
3,54460,10011,105331,232110.0,carevue,admit,None,MICU,NaN,15.0,2126-08-14 22:34:00,2126-08-28 18:59:00,332.42
4,54461,10011,105331,NaN,carevue,discharge,MICU,None,15.0,NaN,2126-08-28 18:59:00,None,NaN


In [36]:
# Column info for every table
tables = conn.execute(
    "SELECT name FROM sqlite_master WHERE type='table' ORDER BY name"
).fetchall()

for (t,) in tables:
    print(f"\n{'='*60}")
    print(f"  {t.upper()}")
    print(f"{'='*60}")
    display(pd.read_sql(f"PRAGMA table_info({t})", conn))


  ADMISSIONS


,cid,name,type,notnull,dflt_value,pk
0,0,row_id,INTEGER,0,None,0
1,1,subject_id,INTEGER,0,None,0
2,2,hadm_id,INTEGER,0,None,0
3,3,admittime,TEXT,0,None,0
4,4,dischtime,TEXT,0,None,0
5,5,deathtime,TEXT,0,None,0
6,6,admission_type,TEXT,0,None,0
7,7,admission_location,TEXT,0,None,0
8,8,discharge_location,TEXT,0,None,0
9,9,insurance,TEXT,0,None,0



  CALLOUT


,cid,name,type,notnull,dflt_value,pk
0,0,row_id,INTEGER,0,None,0
1,1,subject_id,INTEGER,0,None,0
2,2,hadm_id,INTEGER,0,None,0
3,3,submit_wardid,INTEGER,0,None,0
4,4,submit_careunit,TEXT,0,None,0
5,5,curr_wardid,INTEGER,0,None,0
6,6,curr_careunit,TEXT,0,None,0
7,7,callout_wardid,INTEGER,0,None,0
8,8,callout_service,TEXT,0,None,0
9,9,request_tele,INTEGER,0,None,0



  CAREGIVERS


,cid,name,type,notnull,dflt_value,pk
0,0,row_id,INTEGER,0,None,0
1,1,cgid,INTEGER,0,None,0
2,2,label,TEXT,0,None,0
3,3,description,TEXT,0,None,0



  CHARTEVENTS


,cid,name,type,notnull,dflt_value,pk
0,0,row_id,INTEGER,0,None,0
1,1,subject_id,INTEGER,0,None,0
2,2,hadm_id,INTEGER,0,None,0
3,3,icustay_id,REAL,0,None,0
4,4,itemid,INTEGER,0,None,0
5,5,charttime,TEXT,0,None,0
6,6,storetime,TEXT,0,None,0
7,7,cgid,INTEGER,0,None,0
8,8,value,REAL,0,None,0
9,9,valuenum,REAL,0,None,0



  CPTEVENTS


,cid,name,type,notnull,dflt_value,pk
0,0,row_id,INTEGER,0,None,0
1,1,subject_id,INTEGER,0,None,0
2,2,hadm_id,INTEGER,0,None,0
3,3,costcenter,TEXT,0,None,0
4,4,chartdate,TEXT,0,None,0
5,5,cpt_cd,INTEGER,0,None,0
6,6,cpt_number,INTEGER,0,None,0
7,7,cpt_suffix,REAL,0,None,0
8,8,ticket_id_seq,REAL,0,None,0
9,9,sectionheader,TEXT,0,None,0



  D_ICD_DIAGNOSES


,cid,name,type,notnull,dflt_value,pk
0,0,row_id,INTEGER,0,None,0
1,1,icd9_code,TEXT,0,None,0
2,2,short_title,TEXT,0,None,0
3,3,long_title,TEXT,0,None,0



  D_ICD_PROCEDURES


,cid,name,type,notnull,dflt_value,pk
0,0,row_id,INTEGER,0,None,0
1,1,icd9_code,INTEGER,0,None,0
2,2,short_title,TEXT,0,None,0
3,3,long_title,TEXT,0,None,0



  D_ITEMS


,cid,name,type,notnull,dflt_value,pk
0,0,row_id,INTEGER,0,None,0
1,1,itemid,INTEGER,0,None,0
2,2,label,TEXT,0,None,0
3,3,abbreviation,TEXT,0,None,0
4,4,dbsource,TEXT,0,None,0
5,5,linksto,TEXT,0,None,0
6,6,category,TEXT,0,None,0
7,7,unitname,TEXT,0,None,0
8,8,param_type,TEXT,0,None,0
9,9,conceptid,REAL,0,None,0



  D_LABITEMS


,cid,name,type,notnull,dflt_value,pk
0,0,row_id,INTEGER,0,None,0
1,1,itemid,INTEGER,0,None,0
2,2,label,TEXT,0,None,0
3,3,fluid,TEXT,0,None,0
4,4,category,TEXT,0,None,0
5,5,loinc_code,TEXT,0,None,0



  DATETIMEEVENTS


,cid,name,type,notnull,dflt_value,pk
0,0,row_id,INTEGER,0,None,0
1,1,subject_id,INTEGER,0,None,0
2,2,hadm_id,INTEGER,0,None,0
3,3,icustay_id,REAL,0,None,0
4,4,itemid,INTEGER,0,None,0
5,5,charttime,TEXT,0,None,0
6,6,storetime,TEXT,0,None,0
7,7,cgid,INTEGER,0,None,0
8,8,value,TEXT,0,None,0
9,9,valueuom,TEXT,0,None,0



  DIAGNOSES_ICD


,cid,name,type,notnull,dflt_value,pk
0,0,row_id,INTEGER,0,None,0
1,1,subject_id,INTEGER,0,None,0
2,2,hadm_id,INTEGER,0,None,0
3,3,seq_num,INTEGER,0,None,0
4,4,icd9_code,TEXT,0,None,0



  DRGCODES


,cid,name,type,notnull,dflt_value,pk
0,0,row_id,INTEGER,0,None,0
1,1,subject_id,INTEGER,0,None,0
2,2,hadm_id,INTEGER,0,None,0
3,3,drg_type,TEXT,0,None,0
4,4,drg_code,INTEGER,0,None,0
5,5,description,TEXT,0,None,0
6,6,drg_severity,REAL,0,None,0
7,7,drg_mortality,REAL,0,None,0



  ICUSTAYS


,cid,name,type,notnull,dflt_value,pk
0,0,row_id,INTEGER,0,None,0
1,1,subject_id,INTEGER,0,None,0
2,2,hadm_id,INTEGER,0,None,0
3,3,icustay_id,INTEGER,0,None,0
4,4,dbsource,TEXT,0,None,0
5,5,first_careunit,TEXT,0,None,0
6,6,last_careunit,TEXT,0,None,0
7,7,first_wardid,INTEGER,0,None,0
8,8,last_wardid,INTEGER,0,None,0
9,9,intime,TEXT,0,None,0



  INPUTEVENTS_CV


,cid,name,type,notnull,dflt_value,pk
0,0,row_id,INTEGER,0,None,0
1,1,subject_id,INTEGER,0,None,0
2,2,hadm_id,INTEGER,0,None,0
3,3,icustay_id,INTEGER,0,None,0
4,4,charttime,TEXT,0,None,0
5,5,itemid,INTEGER,0,None,0
6,6,amount,REAL,0,None,0
7,7,amountuom,TEXT,0,None,0
8,8,rate,REAL,0,None,0
9,9,rateuom,TEXT,0,None,0



  INPUTEVENTS_MV


,cid,name,type,notnull,dflt_value,pk
0,0,row_id,INTEGER,0,None,0
1,1,subject_id,INTEGER,0,None,0
2,2,hadm_id,INTEGER,0,None,0
3,3,icustay_id,INTEGER,0,None,0
4,4,starttime,TEXT,0,None,0
5,5,endtime,TEXT,0,None,0
6,6,itemid,INTEGER,0,None,0
7,7,amount,REAL,0,None,0
8,8,amountuom,TEXT,0,None,0
9,9,rate,REAL,0,None,0



  LABEVENTS


,cid,name,type,notnull,dflt_value,pk
0,0,row_id,INTEGER,0,None,0
1,1,subject_id,INTEGER,0,None,0
2,2,hadm_id,REAL,0,None,0
3,3,itemid,INTEGER,0,None,0
4,4,charttime,TEXT,0,None,0
5,5,value,TEXT,0,None,0
6,6,valuenum,REAL,0,None,0
7,7,valueuom,TEXT,0,None,0
8,8,flag,TEXT,0,None,0



  MICROBIOLOGYEVENTS


,cid,name,type,notnull,dflt_value,pk
0,0,row_id,INTEGER,0,None,0
1,1,subject_id,INTEGER,0,None,0
2,2,hadm_id,INTEGER,0,None,0
3,3,chartdate,TEXT,0,None,0
4,4,charttime,TEXT,0,None,0
5,5,spec_itemid,INTEGER,0,None,0
6,6,spec_type_desc,TEXT,0,None,0
7,7,org_itemid,REAL,0,None,0
8,8,org_name,TEXT,0,None,0
9,9,isolate_num,REAL,0,None,0



  OUTPUTEVENTS


,cid,name,type,notnull,dflt_value,pk
0,0,row_id,INTEGER,0,None,0
1,1,subject_id,INTEGER,0,None,0
2,2,hadm_id,INTEGER,0,None,0
3,3,icustay_id,REAL,0,None,0
4,4,charttime,TEXT,0,None,0
5,5,itemid,INTEGER,0,None,0
6,6,value,REAL,0,None,0
7,7,valueuom,TEXT,0,None,0
8,8,storetime,TEXT,0,None,0
9,9,cgid,INTEGER,0,None,0



  PATIENTS


,cid,name,type,notnull,dflt_value,pk
0,0,row_id,INTEGER,0,None,0
1,1,subject_id,INTEGER,0,None,0
2,2,gender,TEXT,0,None,0
3,3,dob,TEXT,0,None,0
4,4,dod,TEXT,0,None,0
5,5,dod_hosp,TEXT,0,None,0
6,6,dod_ssn,TEXT,0,None,0
7,7,expire_flag,INTEGER,0,None,0



  PRESCRIPTIONS


,cid,name,type,notnull,dflt_value,pk
0,0,row_id,INTEGER,0,None,0
1,1,subject_id,INTEGER,0,None,0
2,2,hadm_id,INTEGER,0,None,0
3,3,icustay_id,REAL,0,None,0
4,4,startdate,TEXT,0,None,0
5,5,enddate,TEXT,0,None,0
6,6,drug_type,TEXT,0,None,0
7,7,drug,TEXT,0,None,0
8,8,drug_name_poe,TEXT,0,None,0
9,9,drug_name_generic,TEXT,0,None,0



  PROCEDURES_ICD


,cid,name,type,notnull,dflt_value,pk
0,0,row_id,INTEGER,0,None,0
1,1,subject_id,INTEGER,0,None,0
2,2,hadm_id,INTEGER,0,None,0
3,3,seq_num,INTEGER,0,None,0
4,4,icd9_code,INTEGER,0,None,0



  SERVICES


,cid,name,type,notnull,dflt_value,pk
0,0,row_id,INTEGER,0,None,0
1,1,subject_id,INTEGER,0,None,0
2,2,hadm_id,INTEGER,0,None,0
3,3,transfertime,TEXT,0,None,0
4,4,prev_service,TEXT,0,None,0
5,5,curr_service,TEXT,0,None,0



  TRANSFERS


,cid,name,type,notnull,dflt_value,pk
0,0,row_id,INTEGER,0,None,0
1,1,subject_id,INTEGER,0,None,0
2,2,hadm_id,INTEGER,0,None,0
3,3,icustay_id,REAL,0,None,0
4,4,dbsource,TEXT,0,None,0
5,5,eventtype,TEXT,0,None,0
6,6,prev_careunit,TEXT,0,None,0
7,7,curr_careunit,TEXT,0,None,0
8,8,prev_wardid,REAL,0,None,0
9,9,curr_wardid,REAL,0,None,0


In [37]:
# How many unique patients?
display(pd.read_sql("SELECT COUNT(DISTINCT subject_id) AS unique_patients FROM icustays", conn))

# How many unique admissions?
display(pd.read_sql("SELECT COUNT(DISTINCT hadm_id) AS unique_admissions FROM icustays", conn))

# How many total ICU stays?
display(pd.read_sql("SELECT COUNT(*) AS total_icu_stays FROM icustays", conn))

# Patients with more than one ICU stay
display(pd.read_sql("""
    SELECT subject_id, COUNT(*) AS n_stays
    FROM icustays
    GROUP BY subject_id
    HAVING n_stays > 1
    ORDER BY n_stays DESC
""", conn))

,unique_patients
0,100


,unique_admissions
0,129


,total_icu_stays
0,136


,subject_id,n_stays
0,41976,15
1,10124,4
2,10088,3
3,44083,3
4,10059,2
5,10094,2
6,10117,2
7,10119,2
8,40124,2
9,40177,2


# SQL EDA aggregations

## Table View 1: v_cohort

Aggregation to view 
1) ICU Length of Stay 
2) Hospital Length of Stay
3) Age at admission 

In [6]:
# ========================================
# Table View 1: v_cohort
# =========================================
# Can join icustays with admission through 'hadm_id'
# Can join icustays with patients through 'subject_id'

# From icustays need columns: [intime, outtime] 
#   To determine length of stay in ICU
# FROM admissions need columns: [admittime, dischtime]
#   To determine hospital length of stay
# FROM patients need columns: [dob]
#   determine the age of patients using admittime
pd.read_sql(
    """
    SELECT 
        p.subject_id, p.dob,
        i.intime, i.outtime, 
        a.admittime, a.dischtime,
        
        ROUND((JULIANDAY(outtime) - JULIANDAY(intime)) * 24, 2) AS icu_los_hours,
        ROUND(JULIANDAY(dischtime) - JULIANDAY(admittime), 2) AS hosp_los_days,
        ROUND((JULIANDAY(admittime) - JULIANDAY(dob)) / 365, 1) AS age_at_admit
    FROM icustays AS i 
    JOIN admissions AS a ON i.hadm_id = a.hadm_id
    JOIN patients AS p ON a.subject_id = p.subject_id
    """, 
    conn
)
# Note: There are some people about 300 years old. This is okay since this data was modified by PhysioNet to be untraceable/deidentified.
# The SQL query correctly calculates the difference for DOB

,subject_id,dob,intime,outtime,admittime,dischtime,icu_los_hours,hosp_los_days,age_at_admit
0,10006,2094-03-05,2164-10-23 21:10:15,2164-10-25 12:21:07,2164-10-23 21:09:00,2164-11-01 17:15:00,39.18,8.84,70.7
1,10011,2090-06-05,2126-08-14 22:34:00,2126-08-28 18:59:00,2126-08-14 22:32:00,2126-08-28 18:59:00,332.42,13.85,36.2
2,10013,2038-09-03,2125-10-04 23:38:00,2125-10-07 15:13:52,2125-10-04 23:36:00,2125-10-07 15:13:00,63.60,2.65,87.1
3,10017,2075-09-21,2149-05-29 18:52:29,2149-05-31 22:19:17,2149-05-26 17:19:00,2149-06-03 18:42:00,51.45,8.06,73.7
4,10019,2114-06-20,2163-05-14 20:43:56,2163-05-16 03:47:04,2163-05-14 20:43:00,2163-05-15 12:00:00,31.05,0.64,48.9
5,10026,1895-05-17,2195-05-17 07:40:18,2195-05-19 17:24:25,2195-05-17 07:39:00,2195-05-24 11:45:00,57.74,7.17,300.2
6,10027,2108-01-15,2190-07-13 10:39:07,2190-07-20 16:55:45,2190-07-13 07:15:00,2190-07-25 14:00:00,174.28,12.28,82.5
7,10029,2061-04-10,2139-09-23 12:37:10,2139-09-25 18:55:04,2139-09-22 10:58:00,2139-10-02 14:29:00,54.30,10.15,78.5
8,10032,2050-03-29,2138-04-08 01:38:18,2138-04-11 18:43:34,2138-04-02 19:52:00,2138-04-15 14:35:00,89.09,12.78,88.1
9,10033,2051-04-21,2132-12-05 13:13:18,2132-12-06 14:53:22,2132-12-05 02:46:00,2132-12-08 15:15:00,25.67,3.52,81.7


## Table View 2: v_vitals_24h

In order to create Table View 2 we need to know what each itemid is. This is stored in d_items. 

In [53]:
pd.read_sql("""
    SELECT itemid, label, category, unitname, dbsource
    FROM d_items
    WHERE label LIKE '%heart rate%'
        OR label LIKE '%blood pressure%'
        OR label LIKE '%SpO2%'
        OR label LIKE '%oxygen%'
        OR label LIKE '%temperature%'
        OR label LIKE '%GCS%'
        OR label LIKE '%respiratory%'
    ORDER BY dbsource, label
""", conn)

,itemid,label,category,unitname,dbsource
0,198,GCS Total,None,None,carevue
1,211,Heart Rate,None,None,carevue
2,3494,Lowest Heart Rate,None,None,carevue
3,7146,Oxygen #3,None,None,carevue
4,591,RLE [Temperature],None,None,carevue
5,597,RUE [Temperature],None,None,carevue
6,616,Respiratory Effort,None,None,carevue
7,617,Respiratory Pattern,None,None,carevue
8,618,Respiratory Rate,None,None,carevue
9,619,Respiratory Rate Set,None,None,carevue


Further refine the category column to view 'routine vital signs'

In [60]:
pd.read_sql("""
    SELECT itemid, label, category, unitname, dbsource
    FROM d_items
    WHERE category LIKE '%Routine%'
    ORDER BY dbsource, label
""", conn)

,itemid,label,category,unitname,dbsource
0,225310,ART BP Diastolic,Routine Vital Signs,mmHg,metavision
1,225309,ART BP Systolic,Routine Vital Signs,mmHg,metavision
2,225312,ART BP mean,Routine Vital Signs,mmHg,metavision
3,227630,Arctic Sun Temp #1 Location,Routine Vital Signs,None,metavision
4,227631,Arctic Sun Temp #2 Location,Routine Vital Signs,None,metavision
5,227632,Arctic Sun/Alsius Temp #1 C,Routine Vital Signs,?C,metavision
6,227634,Arctic Sun/Alsius Temp #2 C,Routine Vital Signs,?C,metavision
7,220051,Arterial Blood Pressure diastolic,Routine Vital Signs,mmHg,metavision
8,220052,Arterial Blood Pressure mean,Routine Vital Signs,mmHg,metavision
9,220050,Arterial Blood Pressure systolic,Routine Vital Signs,mmHg,metavision


In [55]:
pd.read_sql("""
    SELECT itemid, label, category, unitname
    FROM d_items
    WHERE category LIKE '%Routine%'
    AND (
        label LIKE '%heart rate%'
        OR label LIKE '%blood pressure%'
        OR label LIKE '%SpO2%'
        OR label LIKE '%temperature%'
        OR label LIKE '%respiratory%'
        OR label LIKE '%GCS%'
    )
    ORDER BY dbsource, label
""", conn)

,itemid,label,category,unitname
0,220051,Arterial Blood Pressure diastolic,Routine Vital Signs,mmHg
1,220052,Arterial Blood Pressure mean,Routine Vital Signs,mmHg
2,220050,Arterial Blood Pressure systolic,Routine Vital Signs,mmHg
3,226329,Blood Temperature CCO (C),Routine Vital Signs,?C
4,220045,Heart Rate,Routine Vital Signs,bpm
5,224643,Manual Blood Pressure Diastolic Left,Routine Vital Signs,mmHg
6,227242,Manual Blood Pressure Diastolic Right,Routine Vital Signs,mmHg
7,224167,Manual Blood Pressure Systolic Left,Routine Vital Signs,mmHg
8,227243,Manual Blood Pressure Systolic Right,Routine Vital Signs,mmHg
9,220180,Non Invasive Blood Pressure diastolic,Routine Vital Signs,mmHg


### Metavision System: 

| Vital Sign | Item Id |
|:----------:|:-------:|
| Heart Rate | 220045 |
| Systolic BP | 220050 |
| Sp02 | 220277 |
| Respiratory Rate | 220210 | 
| Temperature (C) | 223762 | 
| GCS Total | 223900 |


In [56]:
pd.read_sql("""
    SELECT itemid, label, category, unitname, dbsource
    FROM d_items
    WHERE itemid IN (211, 220045, 51, 220050, 646, 220277, 
                     618, 220210, 676, 223762, 198, 223900)
    ORDER BY dbsource, label
""", conn)

,itemid,label,category,unitname,dbsource
0,51,Arterial BP [Systolic],None,None,carevue
1,198,GCS Total,None,None,carevue
2,211,Heart Rate,None,None,carevue
3,618,Respiratory Rate,None,None,carevue
4,646,SpO2,None,None,carevue
5,676,Temperature C,None,None,carevue
6,220050,Arterial Blood Pressure systolic,Routine Vital Signs,mmHg,metavision
7,223900,GCS - Verbal Response,Neurological,None,metavision
8,220045,Heart Rate,Routine Vital Signs,bpm,metavision
9,220277,O2 saturation pulseoxymetry,Respiratory,%,metavision


Create the `v_vitals_24h` query. 
1. The two tables are `icustays` and `chartevents`
2. Join on `subject_id`
3. Time window is the first 24 hours using DATETIME(intime, '+24hours')
4. The item IDs which were verified and discovered through d_items
5. Aggregation pattern - AVG/MIN/MAX(CASE WHEN itemid IN (...) THEN valuenum END)

In [75]:
pd.read_sql(
    """
    SELECT 
        i.icustay_id,

        -- Heart Rate
        ROUND(AVG(CASE WHEN c.itemid IN (211, 220045) THEN c.valuenum END), 1) AS hr_mean_24h,
        ROUND(MIN(CASE WHEN c.itemid IN (211, 220045) THEN c.valuenum END), 1) AS hr_min_24h,
        ROUND(MAX(CASE WHEN c.itemid IN (211, 220045) THEN c.valuenum END), 1) AS hr_max_24h,

        -- Systolic BP
        ROUND(AVG(CASE WHEN c.itemid IN (51, 220179) THEN c.valuenum END), 1) AS sbp_mean_24h,
        ROUND(MIN(CASE WHEN c.itemid IN (51, 220179) THEN c.valuenum END), 1) AS sbp_min_24h,
        ROUND(MAX(CASE WHEN c.itemid IN (51, 220179) THEN c.valuenum END), 1) AS sbp_max_24h,

        -- SpO2
        ROUND(AVG(CASE WHEN c.itemid IN (646, 220277) THEN c.valuenum END), 1) AS spo2_mean_24h,
        ROUND(MIN(CASE WHEN c.itemid IN (646, 220277) THEN c.valuenum END), 1) AS spo2_min_24h,
        ROUND(MAX(CASE WHEN c.itemid IN (646, 220277) THEN c.valuenum END), 1) AS spo2_max_24h,

        -- Respiratory Rate
        ROUND(AVG(CASE WHEN c.itemid IN (618, 220210) THEN c.valuenum END), 1) AS resp_mean_24h,
        ROUND(MIN(CASE WHEN c.itemid IN (618, 220210) THEN c.valuenum END), 1) AS resp_min_24h,
        ROUND(MAX(CASE WHEN c.itemid IN (618, 220210) THEN c.valuenum END), 1) AS resp_max_24h,

        -- Temperature (Fahrenheit)
        ROUND(AVG(CASE WHEN c.itemid IN (678, 223761) THEN c.valuenum END), 1) AS temp_mean_24h,
        ROUND(MIN(CASE WHEN c.itemid IN (678, 223761) THEN c.valuenum END), 1) AS temp_min_24h,
        ROUND(MAX(CASE WHEN c.itemid IN (678, 223761) THEN c.valuenum END), 1) AS temp_max_24h,

        -- GCS Total
        ROUND(AVG(CASE WHEN c.itemid IN (198, 223900) THEN c.valuenum END), 1) AS gcs_mean_24h,
        ROUND(MIN(CASE WHEN c.itemid IN (198, 223900) THEN c.valuenum END), 1) AS gcs_min_24h,
        ROUND(MAX(CASE WHEN c.itemid IN (198, 223900) THEN c.valuenum END), 1) AS gcs_max_24h,

        -- Total vital measurements
        COUNT(c.valuenum) AS n_vital_measurements_24h

    FROM icustays AS i
    JOIN chartevents AS c ON i.subject_id = c.subject_id
    WHERE c.charttime >= i.intime
    AND   c.charttime <= DATETIME(i.intime, '+24 hours')
    AND   c.valuenum IS NOT NULL
    AND   c.valuenum > 0
    GROUP BY i.icustay_id
    """,
    conn
)

,icustay_id,hr_mean_24h,hr_min_24h,hr_max_24h,sbp_mean_24h,sbp_min_24h,sbp_max_24h,spo2_mean_24h,spo2_min_24h,spo2_max_24h,resp_mean_24h,resp_min_24h,resp_max_24h,temp_mean_24h,temp_min_24h,temp_max_24h,gcs_mean_24h,gcs_min_24h,gcs_max_24h,n_vital_measurements_24h
0,201006,82.7,65.0,122.0,106.3,87.0,127.0,95.0,86.0,100.0,26.6,15.0,39.0,100.6,98.0,103.4,6.4,3.0,15.0,794
1,201204,68.8,57.0,80.0,134.3,93.0,180.0,98.4,89.0,100.0,18.0,12.0,28.0,96.1,95.1,96.8,3.1,2.0,4.0,435
2,203766,55.0,46.0,74.0,108.1,70.0,154.0,95.5,57.0,100.0,14.4,2.0,34.0,96.5,96.3,96.6,9.0,3.0,15.0,1026
3,204132,97.5,80.0,108.0,139.8,128.0,155.0,99.9,99.0,100.0,17.7,13.0,22.0,98.2,97.0,100.1,1.0,1.0,1.0,492
4,204201,81.9,67.0,89.0,NaN,NaN,NaN,99.9,99.0,100.0,14.8,13.0,19.0,96.6,96.3,96.8,10.3,4.0,14.0,207
5,204881,95.2,82.0,104.0,NaN,NaN,NaN,92.8,90.0,96.0,25.9,15.0,32.0,96.1,95.0,97.4,14.3,14.0,15.0,401
6,205170,87.2,75.0,104.0,98.3,92.0,103.0,99.2,96.0,100.0,14.6,11.0,25.0,98.8,96.8,100.0,1.6,1.0,4.0,510
7,205589,69.3,53.0,84.0,130.3,109.0,154.0,98.0,92.0,100.0,22.6,17.0,31.0,97.6,97.0,98.2,13.9,13.0,14.0,733
8,206504,86.1,70.0,104.0,NaN,NaN,NaN,98.5,97.0,100.0,28.9,15.0,91.0,98.6,97.5,99.6,15.0,15.0,15.0,318
9,209797,73.2,61.0,108.0,101.5,68.0,138.0,98.6,93.0,100.0,14.7,9.0,22.0,97.6,97.0,99.1,1.0,1.0,1.0,407


## v_labs_24h

In [76]:
"""
What tables we need: 
    1. `icustays`: the ICU stay ID and `intime`
    2. `labevents`: individual lab results with `itemid`, `charttime`, and `valuenum`
How do they join?:
    Both share subject_id
Time Window?:
    Same as the vitals: 
    WHERE l.charttime >= i.intime
    AND   l.charttime <= DATETIME(i.intime, '+24 hours')
"""

"\nWhat tables we need: \n    1. `icustays`: the ICU stay ID and `intime`\n    2. `labevents`: individual lab results with `itemid`, `charttime`, and `valuenum`\nHow do they join?:\n    Both share subject_id\nTime Window?:\n    Same as the vitals: \n    WHERE l.charttime >= i.intime\n    AND   l.charttime <= DATETIME(i.intime, '+24 hours')\n"

Know what the codes mean 

In [102]:
# View 10 code definitions
pd.read_sql("SELECT * FROM d_icd_diagnoses LIMIT 10", conn)

,row_id,icd9_code,short_title,long_title
0,1,01716,Erythem nod tb-oth test,"Erythema nodosum with hypersensitivity reaction in tuberculosis, tubercle bacilli not found by bacteriological or histological examination, but tuberculosis confirmed by other methods [inoculation of animals]"
1,2,01720,TB periph lymph-unspec,"Tuberculosis of peripheral lymph nodes, unspecified"
2,3,01721,TB periph lymph-no exam,"Tuberculosis of peripheral lymph nodes, bacteriological or histological examination not done"
3,4,01722,TB periph lymph-exam unk,"Tuberculosis of peripheral lymph nodes, bacteriological or histological examination unknown (at present)"
4,5,01723,TB periph lymph-micro dx,"Tuberculosis of peripheral lymph nodes, tubercle bacilli found (in sputum) by microscopy"
5,6,01724,TB periph lymph-cult dx,"Tuberculosis of peripheral lymph nodes, tubercle bacilli not found (in sputum) by microscopy, but found by bacterial culture"
6,7,01725,TB periph lymph-histo dx,"Tuberculosis of peripheral lymph nodes, tubercle bacilli not found by bacteriological examination, but tuberculosis confirmed histologically"
7,8,01726,TB periph lymph-oth test,"Tuberculosis of peripheral lymph nodes, tubercle bacilli not found by bacteriological or histological examination, but tuberculosis confirmed by other methods [inoculation of animals]"
8,9,01730,TB of eye-unspec,"Tuberculosis of eye, unspecified"
9,10,01731,TB of eye-no exam,"Tuberculosis of eye, bacteriological or histological examination not done"


In [81]:
pd.read_sql("""
    SELECT l.itemid, d.label, d.fluid, d.category, COUNT(*) AS n_measurements
    FROM labevents l
    JOIN d_labitems d ON l.itemid = d.itemid
    GROUP BY l.itemid, d.label, d.fluid, d.category
    ORDER BY n_measurements DESC
""", conn)

,itemid,label,fluid,category,n_measurements
0,51221,Hematocrit,Blood,Hematology,2317
1,50971,Potassium,Blood,Chemistry,2279
2,50983,Sodium,Blood,Chemistry,2185
3,50912,Creatinine,Blood,Chemistry,2175
4,50902,Chloride,Blood,Chemistry,2160
5,51006,Urea Nitrogen,Blood,Chemistry,2158
6,50882,Bicarbonate,Blood,Chemistry,2151
7,50868,Anion Gap,Blood,Chemistry,2134
8,50931,Glucose,Blood,Chemistry,2121
9,51265,Platelet Count,Blood,Hematology,2088


There are many lab vitals but as we can see there are some that are simply ordered more often than others. We can choose the most ordered since we can assume they are the most common. 

In [93]:
# Lets view the top 20 lab vitals
print(pd.read_sql("""
    SELECT l.itemid, d.label, d.fluid, d.category, COUNT(*) AS n_measurements
    FROM labevents l
    JOIN d_labitems d ON l.itemid = d.itemid
    GROUP BY l.itemid, d.label, d.fluid, d.category
    ORDER BY n_measurements DESC
    LIMIT 20
""", conn))

    itemid              label  fluid    category  n_measurements
0    51221         Hematocrit  Blood  Hematology            2317
1    50971          Potassium  Blood   Chemistry            2279
2    50983             Sodium  Blood   Chemistry            2185
3    50912         Creatinine  Blood   Chemistry            2175
4    50902           Chloride  Blood   Chemistry            2160
5    51006      Urea Nitrogen  Blood   Chemistry            2158
6    50882        Bicarbonate  Blood   Chemistry            2151
7    50868          Anion Gap  Blood   Chemistry            2134
8    50931            Glucose  Blood   Chemistry            2121
9    51265     Platelet Count  Blood  Hematology            2088
10   51222         Hemoglobin  Blood  Hematology            2024
11   51301  White Blood Cells  Blood  Hematology            2021
12   51249               MCHC  Blood  Hematology            2008
13   51248                MCH  Blood  Hematology            2007
14   51250               

In [ ]:
# In this case there is only one vital id. 
# This makes sense b/c this vital comes from lab not measurements from two medical devices
# v_lab
pd.read_sql(
    """
    SELECT
        i.icustay_id,

        -- Hematocrit
        ROUND(AVG(CASE WHEN l.itemid = 51221 THEN l.valuenum END), 2) AS hematocrit_mean_24h,
        ROUND(MIN(CASE WHEN l.itemid = 51221 THEN l.valuenum END), 2) AS hematocrit_min_24h,

        -- Potassium
        ROUND(AVG(CASE WHEN l.itemid = 50971 THEN l.valuenum END), 2) AS potassium_mean_24h,
        ROUND(MIN(CASE WHEN l.itemid = 50971 THEN l.valuenum END), 2) AS potassium_min_24h,
        ROUND(MAX(CASE WHEN l.itemid = 50971 THEN l.valuenum END), 2) AS potassium_max_24h,

        -- Sodium
        ROUND(AVG(CASE WHEN l.itemid = 50983 THEN l.valuenum END), 2) AS sodium_mean_24h,
        ROUND(MIN(CASE WHEN l.itemid = 50983 THEN l.valuenum END), 2) AS sodium_min_24h,
        ROUND(MAX(CASE WHEN l.itemid = 50983 THEN l.valuenum END), 2) AS sodium_max_24h,

        -- Creatinine
        ROUND(AVG(CASE WHEN l.itemid = 50912 THEN l.valuenum END), 2) AS creatinine_mean_24h,
        ROUND(MAX(CASE WHEN l.itemid = 50912 THEN l.valuenum END), 2) AS creatinine_max_24h,

        -- Chloride
        ROUND(AVG(CASE WHEN l.itemid = 50902 THEN l.valuenum END), 2) AS chloride_mean_24h,

        -- Urea Nitrogen (BUN)
        ROUND(AVG(CASE WHEN l.itemid = 51006 THEN l.valuenum END), 2) AS bun_mean_24h,
        ROUND(MAX(CASE WHEN l.itemid = 51006 THEN l.valuenum END), 2) AS bun_max_24h,

        -- Bicarbonate
        ROUND(AVG(CASE WHEN l.itemid = 50882 THEN l.valuenum END), 2) AS bicarb_mean_24h,
        ROUND(MIN(CASE WHEN l.itemid = 50882 THEN l.valuenum END), 2) AS bicarb_min_24h,

        -- Anion Gap
        ROUND(AVG(CASE WHEN l.itemid = 50868 THEN l.valuenum END), 2) AS anion_gap_mean_24h,
        ROUND(MAX(CASE WHEN l.itemid = 50868 THEN l.valuenum END), 2) AS anion_gap_max_24h,

        -- Glucose
        ROUND(AVG(CASE WHEN l.itemid = 50931 THEN l.valuenum END), 2) AS glucose_mean_24h,
        ROUND(MIN(CASE WHEN l.itemid = 50931 THEN l.valuenum END), 2) AS glucose_min_24h,
        ROUND(MAX(CASE WHEN l.itemid = 50931 THEN l.valuenum END), 2) AS glucose_max_24h,

        -- Platelet Count
        ROUND(AVG(CASE WHEN l.itemid = 51265 THEN l.valuenum END), 2) AS platelets_mean_24h,
        ROUND(MIN(CASE WHEN l.itemid = 51265 THEN l.valuenum END), 2) AS platelets_min_24h,

        -- Hemoglobin
        ROUND(AVG(CASE WHEN l.itemid = 51222 THEN l.valuenum END), 2) AS hemoglobin_mean_24h,
        ROUND(MIN(CASE WHEN l.itemid = 51222 THEN l.valuenum END), 2) AS hemoglobin_min_24h,

        -- White Blood Cells
        ROUND(AVG(CASE WHEN l.itemid = 51301 THEN l.valuenum END), 2) AS wbc_mean_24h,
        ROUND(MAX(CASE WHEN l.itemid = 51301 THEN l.valuenum END), 2) AS wbc_max_24h,

        -- MCHC
        ROUND(AVG(CASE WHEN l.itemid = 51249 THEN l.valuenum END), 2) AS mchc_mean_24h,

        -- MCH
        ROUND(AVG(CASE WHEN l.itemid = 51248 THEN l.valuenum END), 2) AS mch_mean_24h,

        -- MCV
        ROUND(AVG(CASE WHEN l.itemid = 51250 THEN l.valuenum END), 2) AS mcv_mean_24h,

        -- Red Blood Cells
        ROUND(AVG(CASE WHEN l.itemid = 51279 THEN l.valuenum END), 2) AS rbc_mean_24h,
        ROUND(MIN(CASE WHEN l.itemid = 51279 THEN l.valuenum END), 2) AS rbc_min_24h,

        -- RDW
        ROUND(AVG(CASE WHEN l.itemid = 51277 THEN l.valuenum END), 2) AS rdw_mean_24h,
        ROUND(MAX(CASE WHEN l.itemid = 51277 THEN l.valuenum END), 2) AS rdw_max_24h,

        -- Magnesium
        ROUND(AVG(CASE WHEN l.itemid = 50960 THEN l.valuenum END), 2) AS magnesium_mean_24h,
        ROUND(MIN(CASE WHEN l.itemid = 50960 THEN l.valuenum END), 2) AS magnesium_min_24h,

        -- Calcium Total
        ROUND(AVG(CASE WHEN l.itemid = 50893 THEN l.valuenum END), 2) AS calcium_mean_24h,
        ROUND(MIN(CASE WHEN l.itemid = 50893 THEN l.valuenum END), 2) AS calcium_min_24h,

        -- Phosphate
        ROUND(AVG(CASE WHEN l.itemid = 50970 THEN l.valuenum END), 2) AS phosphate_mean_24h,
        ROUND(MAX(CASE WHEN l.itemid = 50970 THEN l.valuenum END), 2) AS phosphate_max_24h,

        -- Total lab draws 
        COUNT(*) AS n_lab_draws_24h

    FROM icustays AS i
    JOIN labevents AS l ON i.subject_id = l.subject_id
    WHERE l.charttime >= i.intime
    AND   l.charttime <= DATETIME(i.intime, '+24 hours')
    AND   l.valuenum IS NOT NULL
    AND   l.valuenum > 0
    GROUP BY i.icustay_id
    """,
    conn
)

,icustay_id,hematocrit_mean_24h,hematocrit_min_24h,potassium_mean_24h,potassium_min_24h,potassium_max_24h,sodium_mean_24h,sodium_min_24h,sodium_max_24h,creatinine_mean_24h,...,rbc_min_24h,rdw_mean_24h,rdw_max_24h,magnesium_mean_24h,magnesium_min_24h,calcium_mean_24h,calcium_min_24h,phosphate_mean_24h,phosphate_max_24h,n_lab_draws_24h
0,201006,26.00,25.3,3.60,3.6,3.6,131.00,131.0,131.0,0.80,...,3.09,15.90,15.9,1.80,1.8,7.50,7.5,2.50,2.5,57
1,201204,26.20,24.8,4.25,4.0,4.5,144.00,143.0,145.0,2.45,...,3.26,14.90,14.9,2.15,2.1,8.30,8.3,5.70,5.7,46
2,203766,36.25,34.0,4.50,4.5,4.5,131.00,129.0,133.0,1.55,...,3.90,16.45,17.0,2.75,2.2,7.00,6.9,5.85,6.2,184
3,204132,32.20,31.2,3.55,3.2,3.9,141.00,140.0,142.0,0.40,...,3.66,15.10,15.2,1.95,1.9,7.70,7.4,1.50,2.2,68
4,204201,25.83,21.8,4.12,2.9,5.2,136.40,128.0,143.0,0.53,...,2.63,13.95,14.1,1.93,1.3,4.85,4.7,1.95,2.0,95
5,204881,27.50,27.5,3.00,3.0,3.0,139.00,139.0,139.0,0.30,...,3.15,13.70,13.7,2.20,2.2,7.50,7.5,1.60,1.6,30
6,205170,30.60,30.6,3.10,3.1,3.1,144.00,144.0,144.0,0.50,...,4.23,14.90,14.9,2.10,2.1,6.50,6.5,1.60,1.6,52
7,205589,29.78,27.4,3.95,3.8,4.1,134.50,134.0,135.0,1.05,...,3.28,14.20,14.2,2.10,1.9,8.10,8.0,1.95,2.0,48
8,206504,36.90,36.9,4.40,4.4,4.4,139.00,139.0,139.0,3.50,...,3.77,19.80,19.8,1.40,1.4,8.50,8.5,5.00,5.0,27
9,209797,35.80,35.8,3.30,3.3,3.3,147.50,146.0,149.0,0.75,...,4.85,16.80,16.8,1.90,1.9,7.35,7.2,2.40,2.8,48


Reference list for lab vital aggregation choices

1. Bellomo, R., Ronco, C., Kellum, J. A., Mehta, R. L., & Palevsky, P. (2004). Acute renal failure — definition, outcome measures, animal models, fluid therapy and information technology needs: the Second International Consensus Conference of the Acute Dialysis Quality Initiative (ADQI) Group. *Critical Care*, 8(4), R204–R212.

2. Mehta, R. L., Kellum, J. A., Shah, S. V., Molitoris, B. A., Ronco, C., Warnock, D. G., & Levin, A. (2007). Acute Kidney Injury Network: report of an initiative to improve outcomes in acute kidney injury. *Critical Care*, 11(2), R31.

3. Martini, W. Z., Dubick, M. A., Pusateri, A. E., Park, M. S., Ryan, K. L., & Holcomb, J. B. (2006). Does bicarbonate correct coagulopathy caused by acidosis in swine? *Journal of Trauma*, 61(1), 99–106.

4. Hébert, P. C., Wells, G., Blajchman, M. A., Marshall, J., Martin, C., Pagliarello, G., Tweeddale, M., Schweitzer, I., & Yetisir, E. (1999). A multicenter, randomized, controlled clinical trial of transfusion requirements in critical care. *New England Journal of Medicine*, 340(6), 409–417.

5. Bone, R. C., Balk, R. A., Cerra, F. B., Dellinger, R. P., Fein, A. M., Knaus, W. A., Schein, R. M., & Sibbald, W. J. (1992). Definitions for sepsis and organ failure and guidelines for the use of innovative therapies in sepsis. *Chest*, 101(6), 1644–1655.

6. Knaus, W. A., Draper, E. A., Wagner, D. P., & Zimmerman, J. E. (1985). APACHE II: a severity of disease classification system. *Critical Care Medicine*, 13(10), 818–829.

7. Egi, M., Bellomo, R., Stachowski, E., French, C. J., & Hart, G. (2006). Variability of blood glucose concentration and short-term mortality in critically ill patients. *Anesthesiology*, 105(2), 244–252.

8. Macdonald, J. E., & Struthers, A. D. (2004). What is the optimal serum potassium level in cardiovascular patients? *Heart*, 90(10), 1041–1045.

9. Le Gall, J. R., Lemeshow, S., & Saulnier, F. (1993). A new Simplified Acute Physiology Score (SAPS II) based on a European/North American multicenter study. *JAMA*, 270(24), 2957–2963.

10. Vincent, J. L., Moreno, R., Takala, J., Willatts, S., De Mendonça, A., Bruining, H., Reinhart, C. K., Suter, P. M., & Thijs, L. G. (1996). The SOFA (Sepsis-related Organ Failure Assessment) score to describe organ dysfunction/failure. *Intensive Care Medicine*, 22(7), 707–710.

## v_comorbidities

In [ ]:
""""
What tables do we need?:
    Need `diagnoses_icd`. Diagnoses are the the admission level. 
    Need `d_icd_diagnoses`. Lets us know what the diagnosis is.
    
Purpose: 
    Identify which diagnoses the patient was identified with 
"""

'"\nWhat tables do we need?:\n    Only need `diagnoses_icd`. Diagnoses are the the admission level. \n\nPurpose: \n    Identify which diagnoses the patient was identified with \n'

In [107]:
# Determine top 20 diagnoses
print(pd.read_sql(
    """
    SELECT 
        d.icd9_code,
        di.short_title,
        di.long_title,
        COUNT(*) AS n
    FROM diagnoses_icd AS d
    JOIN d_icd_diagnoses AS di ON d.icd9_code = di.icd9_code
    GROUP BY d.icd9_code, di.short_title, di.long_title
    ORDER BY n DESC
    LIMIT 20
    """,
    conn
))

   icd9_code               short_title  \
0       4019          Hypertension NOS   
1      42731       Atrial fibrillation   
2       5849  Acute kidney failure NOS   
3       4280                   CHF NOS   
4      25000  DMII wo cmp nt st uncntr   
5      51881  Acute respiratry failure   
6       2724    Hyperlipidemia NEC/NOS   
7       5990  Urin tract infection NOS   
8        486   Pneumonia, organism NOS   
9       2859                Anemia NOS   
10     99592             Severe sepsis   
11      0389            Septicemia NOS   
12      2449        Hypothyroidism NOS   
13     41401  Crnry athrscl natve vssl   
14     V5861  Long-term use anticoagul   
15      2760           Hyperosmolality   
16      2761             Hyposmolality   
17      2762                  Acidosis   
18       311   Depressive disorder NEC   
19      5070    Food/vomit pneumonitis   

                                                                                                    long_title  \
0  

In [108]:
pd.read_sql("""
SELECT
d.subject_id,
d.hadm_id,

-- Hypertension
MAX(CASE WHEN d.icd9_code = '4019' THEN 1 ELSE 0 END) AS has_hypertension,

-- Atrial Fibrillation
MAX(CASE WHEN d.icd9_code = '42731' THEN 1 ELSE 0 END) AS has_afib,

-- Acute Kidney Failure
MAX(CASE WHEN d.icd9_code LIKE '584%' THEN 1 ELSE 0 END) AS has_aki,

-- CHF
MAX(CASE WHEN d.icd9_code LIKE '428%' THEN 1 ELSE 0 END) AS has_chf,

-- Diabetes
MAX(CASE WHEN d.icd9_code LIKE '250%' THEN 1 ELSE 0 END) AS has_diabetes,

-- Acute Respiratory Failure
MAX(CASE WHEN d.icd9_code = '51881' THEN 1 ELSE 0 END) AS has_resp_failure,

-- Sepsis (both codes combined)
MAX(CASE WHEN d.icd9_code IN ('99592', '0389') THEN 1 ELSE 0 END) AS has_sepsis,

-- Anemia
MAX(CASE WHEN d.icd9_code LIKE '285%' THEN 1 ELSE 0 END) AS has_anemia,

-- Coronary Artery Disease
MAX(CASE WHEN d.icd9_code LIKE '414%' THEN 1 ELSE 0 END) AS has_cad,

-- Acidosis
MAX(CASE WHEN d.icd9_code = '2762' THEN 1 ELSE 0 END) AS has_acidosis,

-- Total diagnoses (complexity proxy)
COUNT(DISTINCT d.icd9_code) AS n_diagnoses

FROM diagnoses_icd AS d
GROUP BY d.subject_id, d.hadm_id;
""", conn)

,subject_id,hadm_id,has_hypertension,has_afib,has_aki,has_chf,has_diabetes,has_resp_failure,has_sepsis,has_anemia,has_cad,has_acidosis,n_diagnoses
0,10006,142345,0,1,0,1,1,0,0,1,1,0,21
1,10011,105331,0,0,0,0,0,0,0,0,0,0,6
2,10013,165520,0,1,0,0,0,0,1,0,0,0,9
3,10017,199207,1,0,0,0,1,0,0,1,0,0,14
4,10019,177759,1,0,1,0,0,1,1,0,0,0,14
5,10026,103770,0,0,0,0,0,0,0,1,0,0,7
6,10027,199395,1,1,0,1,1,0,0,0,1,0,12
7,10029,132349,0,1,0,1,0,0,1,0,0,0,23
8,10032,140372,0,0,1,1,0,0,0,1,0,0,9
9,10033,157235,1,1,1,0,1,0,0,0,0,0,12


## v_prior_admissions 

In [ ]:
"""
What table do we need?:
    1. admissions

Purpose: 
    Count how many times each patient was admitted before the current admission.
"""

In [ ]:
# See what patients have many admissions
pd.read_sql("""
    SELECT subject_id, COUNT(*) AS n_admissions
    FROM admissions
    GROUP BY subject_id
    HAVING n_admissions > 1
    ORDER BY n_admissions DESC
""", conn)

,subject_id,n_admissions
0,41976,15
1,10088,3
2,44083,3
3,10059,2
4,10094,2
5,10117,2
6,10119,2
7,10124,2
8,40124,2
9,40310,2


A patient can have more than one admission. That is important to know to count only the prior admissions. 
For example 41976 actually should have 14 prior admissions. 

In [116]:
pd.read_sql("""
    SELECT
        a1.subject_id,
        a1.hadm_id,
        COUNT(a2.hadm_id) AS n_prior_admissions
    FROM admissions AS a1
    LEFT JOIN admissions AS a2
        ON  a1.subject_id = a2.subject_id
        AND a2.admittime  < a1.admittime
    GROUP BY a1.subject_id, a1.hadm_id
    ORDER BY n_prior_admissions DESC
""", conn)

,subject_id,hadm_id,n_prior_admissions
0,41976,153826,14
1,41976,149469,13
2,41976,145024,12
3,41976,151798,11
4,41976,179418,10
5,41976,155297,9
6,41976,125013,8
7,41976,174863,7
8,41976,180546,6
9,41976,130681,5


## v_vasopressors

In [117]:
"""
A binary flag for whether vasopressors were used during each ICU stay. 
Given to patients to raise blood pressure when it drops dangerously low. 

What tables do we need?:
    1. inputevents_cv which stores CareVue system 
    2. inputevents_mv which stores Metavisions system
    3. drgcodes which lets us know the drug

"""

'\nA binary flag for whether vasopressors were used during each ICU stay. \nGiven to patients to raise blood pressure when it drops dangerously low. \n\nWhat tables do we need?:\n    1. inputevents_cv which stores CareVue system \n    2. inputevents_mv which stores Metavisions system\n    3. drgcodes which lets us know the drug\n\n'

In [122]:
# View the drug codes 
print(pd.read_sql("""
    SELECT itemid, label, category
    FROM d_items
    WHERE label LIKE '%dopamine%'
    OR    label LIKE '%norepinephrine%'
    OR    label LIKE '%epinephrine%'
    OR    label LIKE '%dobutamine%'
    OR    label LIKE '%vasopressin%'
    ORDER BY label
""", conn))

    itemid                 label          category
0     4501         DOPAMINE DRIP              None
1     5805  DOPAMINE MICS/KG/MIN              None
2    30042            Dobutamine              None
3   221653            Dobutamine       Medications
4    30306       Dobutamine Drip              None
5     5329              Dopamine              None
6    30043              Dopamine              None
7   221662              Dopamine       Medications
8    30307         Dopamine Drip              None
9    30044           Epinephrine              None
10  221289           Epinephrine       Medications
11   30309      Epinephrine Drip              None
12   30119         Epinephrine-k              None
13  221906        Norepinephrine       Medications
14    1222           VASOPRESSIN              None
15    2765  VASOPRESSIN   UNIT/R              None
16   42802   VASOPRESSIN  CC/HR.  Free Form Intake
17    2561      VASOPRESSIN U/HR              None
18    2248  VASOPRESSIN UNIT/MI

There is a clear split that one log is from Carevue and the other form Metavision

In [ ]:
pd.read_sql("""
    SELECT DISTINCT icustay_id, 1 AS vasopressor_flag
    FROM inputevents_cv
    WHERE itemid IN (
        4501, 5805, 30042, 30306, 5329, 30043, 30307,  -- Dopamine/Dobutamine CareVue
        30044, 30309, 30119,                             -- Epinephrine CareVue
        1222, 2765, 42802, 2561, 2248, 6255, 2445,      -- Vasopressin CareVue
        30051, 7341, 5747, 3112, 1136, 2334, 1327,      -- Others CareVue
        42273
    )
    UNION
    SELECT DISTINCT icustay_id, 1 AS vasopressor_flag
    FROM inputevents_mv
    WHERE itemid IN (
        221653,  -- Dobutamine Metavision
        221662,  -- Dopamine Metavision
        221289,  -- Epinephrine Metavision
        221906,  -- Norepinephrine Metavision
        222315   -- Vasopressin Metavision
    )
""", conn)

,icustay_id,vasopressor_flag
0,201006,1
1,203766,1
2,205170,1
3,209797,1
4,210989,1
5,213315,1
6,216185,1
7,217724,1
8,217992,1
9,221684,1


## v_features

With all the tables now settled we can just merge all of them into a master view

In [4]:
with open('../sql/features.sql', 'r') as f:
    sql = f.read()

conn.executescript(sql)
print("All views created")

OperationalError: unrecognized token: "#"

In [5]:
# Read and execute the entire features.sql file
with open('../sql/features.sql', 'r') as f:
    sql = f.read()

# Step 2 — Split on semicolons and execute each statement
for stmt in [s.strip() for s in sql.split(';') if s.strip()]:
    try:
        conn.execute(stmt)
    except Exception as e:
        print(f"Warning: {e}")

print("All views registered")

# Step 3 — Query the master view
df = pd.read_sql("SELECT * FROM v_features", conn)
df

All views registered


,subject_id,hadm_id,icustay_id,gender,admission_type,insurance,ethnicity,dob,dod,intime,...,has_chf,has_diabetes,has_resp_failure,has_sepsis,has_anemia,has_cad,has_acidosis,n_diagnoses,n_prior_admissions,vasopressor_flag
0,10006,142345,206504,F,EMERGENCY,Medicare,BLACK/AFRICAN AMERICAN,2094-03-05,2165-08-12,2164-10-23 21:10:15,...,1,1,0,0,1,1,0,21,0,0
1,10011,105331,232110,F,EMERGENCY,Private,UNKNOWN/NOT SPECIFIED,2090-06-05,2126-08-28,2126-08-14 22:34:00,...,0,0,0,0,0,0,0,6,0,0
2,10013,165520,264446,F,EMERGENCY,Medicare,UNKNOWN/NOT SPECIFIED,2038-09-03,2125-10-07,2125-10-04 23:38:00,...,0,0,0,1,0,0,0,9,0,1
3,10017,199207,204881,F,EMERGENCY,Medicare,WHITE,2075-09-21,2152-09-12,2149-05-29 18:52:29,...,0,1,0,0,1,0,0,14,0,0
4,10019,177759,228977,M,EMERGENCY,Medicare,WHITE,2114-06-20,2163-05-15,2163-05-14 20:43:56,...,0,0,1,1,0,0,0,14,0,1
5,10026,103770,277021,F,EMERGENCY,Medicare,WHITE,1895-05-17,2195-11-24,2195-05-17 07:40:18,...,0,0,0,0,1,0,0,7,0,0
6,10027,199395,286020,F,ELECTIVE,Medicare,WHITE,2108-01-15,2190-09-14,2190-07-13 10:39:07,...,1,1,0,0,0,1,0,12,0,0
7,10029,132349,226055,M,EMERGENCY,Medicare,WHITE,2061-04-10,2140-09-21,2139-09-23 12:37:10,...,1,0,0,1,0,0,0,23,0,0
8,10032,140372,267090,M,EMERGENCY,Medicare,WHITE,2050-03-29,2138-05-21,2138-04-08 01:38:18,...,1,0,0,0,1,0,0,9,0,0
9,10033,157235,254543,F,EMERGENCY,Medicare,WHITE,2051-04-21,2133-09-09,2132-12-05 13:13:18,...,0,1,0,0,0,0,0,12,0,0
